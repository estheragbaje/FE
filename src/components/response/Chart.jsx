import React, { Component } from 'react';
import styled from 'styled-components';
import * as V from 'victory';
import {
  white,
  extraSmallSpace,
  mediumSpace3,
  body1,
  body2,
  fadedBlue,
  bodyHero
} from '../../styles/variables';

class Chart extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  render() {
    const { data } = this.props;
    const {
      VictoryBar,
      VictoryChart,
      VictoryTheme,
      VictoryLegend,
      VictoryAxis
    } = V;
    let empty = 0;
    data.forEach((ques, i) => {
      if (ques.feedbacks.length === 0) {
        empty += 1;
      }
    });

    if (empty === data.length) {
      return (
        <div>
          <NoFeedback>No Response for this survey yet</NoFeedback>
        </div>
      );
    }
    return data.map((ques, i) => {
      if (ques.type === 'rating') {
        const { feedbacks } = ques;
        let freq = [];
        freq = [...new Array(10)].map((item, index) => {
          let count = 0;
          feedbacks.forEach((feedback, i) => {
            if (Number(feedback.rating) === index + 1) {
              count += 1;
            }
          });
          return count;
        }); // freq
        const xRange = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
        const processedData = (frequency, rating) =>
          rating.map((rate, i) => ({
            rating: rate,
            freq: frequency[i]
          }));
        const dataPlot = processedData(freq, xRange);

        return (
          <div key={ques.id}>
            <StyledHeader>{ques.question}</StyledHeader>
            <StyledDiv>
              <VictoryChart
                style={{ parent: { maxWidth: '90%' } }}
                // adding the material theme provided with Victory
                theme={VictoryTheme.material}
              >
                <VictoryLegend
                  x={5}
                  y={8}
                  gutter={5}
                  orientation="horizontal"
                  data={[
                    { name: 'x axis - rating', symbol: { fill: '#6bafe8' } },
                    {
                      name: 'y axis - number of votes',
                      symbol: { fill: '#6bafe8' }
                    }
                  ]}
                />
                <VictoryAxis tickValues={[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]} />
                <VictoryAxis
                  dependentAxis
                  tickFormat={x => (Number.isInteger(x) ? x : '')}
                />
                <VictoryBar
                  style={{ data: { fill: '#6bafe8' } }}
                  data={dataPlot}
                  // data accessor for x values
                  x="rating"
                  // data accessor for y values
                  y="freq"
                />
              </VictoryChart>
            </StyledDiv>
          </div>
        );
      }
      return (
        <div key={ques.id}>
          <StyledHeader>{ques.question}</StyledHeader>
          <TextBox>
            {ques.feedbacks.map(feedback => (
              <div key={feedback.id}>
                {feedback.comment && <Paragraph>{feedback.comment}</Paragraph>}
              </div>
            ))}
          </TextBox>
        </div>
      );
    });
  }
}

const StyledHeader = styled.h4`
  margin: ${mediumSpace3} 0 ${extraSmallSpace} 0;
  font-size: ${body1};
`;
const StyledDiv = styled.div`
  border: 1px solid ${fadedBlue};
  display: flex;
  justify-content: center;
  background-color: ${white};
  width: 75%;

  @media (max-width: 767px) {
    width: 100%;
  }
`;
const TextBox = styled.div`
  font-size: ${body2};
  background-color: ${white};
  min-height: ${extraSmallSpace};
  width: 75%;

  @media (max-width: 767px) {
    width: 100%;
  }
`;

const Paragraph = styled.p`
  border: 1px solid ${fadedBlue};
  padding: ${extraSmallSpace} ${extraSmallSpace};
`;
const NoFeedback = styled.div`
  border: 1px solid ${fadedBlue};
  display: flex;
  justify-content: center;
  align-items: center;
  font-size: ${bodyHero};
  background-color: ${white};
  width: 75%;
  height: 30vh;
`;
export default Chart;
