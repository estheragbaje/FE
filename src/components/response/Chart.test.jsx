import React from 'react';
import { shallow, mount, render } from 'enzyme';
import renderer from 'react-test-renderer';
import Chart from './Chart';


const mockProps = {
  data: [
    {
      id: 1,
      type: 'text',
      question: 'What do you say about my overall usefulness in the team for the past 4 months?',
      feedbacks: [
        {
          id: 1,
          rating: '7',
          comment: null,
        },
        {
          id: 2,
          rating: null,
          comment: "It's been quite incredible",
        },
      ],
    },
    {
      id: 2,
      type: 'text',
      question: 'What do you say about my overall usefulness in the team for the past 4 months?',
      feedbacks: [
        {
          id: 1,
          rating: null,
          comment: "It's been quite incredible",
        },
        {
          id: 2,
          rating: 3,
          comment: null,
        },
      ],
    },
  ],
};

let wrapper;
beforeAll(() => {
  // eslint-disable-next-line react/jsx-props-no-spreading
  wrapper = shallow(<Chart {...mockProps} />);
});
describe('<Chart />', () => {
  it('renders the chart component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders NoFeedBack if feedback is empty', () => {
    const nofeedback = renderer.create(
      <Chart {...mockProps} />,
    ).toJSON();
    expect(nofeedback).toMatchSnapshot();
  });
});
