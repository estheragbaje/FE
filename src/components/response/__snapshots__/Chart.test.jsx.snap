// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chart /> renders NoFeedBack if feedback is empty 1`] = `
Array [
  <div>
    <h4
      className="sc-bdVaJa kMUebk"
    >
      What do you say about my overall usefulness in the team for the past 4 months?
    </h4>
    <div
      className="sc-htpNat jhFAkZ"
    >
      <div />
      <div>
        <p
          className="sc-bxivhb kjDdIz"
        >
          It's been quite incredible
        </p>
      </div>
    </div>
  </div>,
  <div>
    <h4
      className="sc-bdVaJa kMUebk"
    >
      What do you say about my overall usefulness in the team for the past 4 months?
    </h4>
    <div
      className="sc-htpNat jhFAkZ"
    >
      <div>
        <p
          className="sc-bxivhb kjDdIz"
        >
          It's been quite incredible
        </p>
      </div>
      <div />
    </div>
  </div>,
]
`;

exports[`<Chart /> renders the chart component correctly 1`] = `
Array [
  <div
    key="1"
  >
    <styled.h4>
      What do you say about my overall usefulness in the team for the past 4 months?
    </styled.h4>
    <styled.div>
      <div
        key="1"
      />
      <div
        key="2"
      >
        <styled.p>
          It's been quite incredible
        </styled.p>
      </div>
    </styled.div>
  </div>,
  <div
    key="2"
  >
    <styled.h4>
      What do you say about my overall usefulness in the team for the past 4 months?
    </styled.h4>
    <styled.div>
      <div
        key="1"
      >
        <styled.p>
          It's been quite incredible
        </styled.p>
      </div>
      <div
        key="2"
      />
    </styled.div>
  </div>,
]
`;
